### PR TITLE
feat(api): manage acknowledgements

### PR DESCRIPTION
**Pull request by callapa**
# Description

Allow to add/list/remove an acknowledgement on host or service

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [ ] 19.04.x
- [x] 19.10.x (master)

#### Community contributors & Centreon team

- [x] I followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

#### Centreon team only

- [ ] I have made sure that the **unit tests** related to the story are successful.
- [ ] I have made sure that **unit tests cover 80%** of the code written for the story.
- [ ] I have made sure that **acceptance tests** related to the story are successful (**local and CI**)

### CODE REVIEW

File: src/Centreon/Domain/Acknowledgement/AcknowledgementService.php [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        $this->engineService = $engineService;
+        $this->validator = $validator;
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param Contact $contact
+     * @return AcknowledgementServiceInterface
+     */
+    public function filterByContact($contact)
```
Comment (kduret) [inline src/Centreon/Domain/Acknowledgement/AcknowledgementService.php]:
```suggestion
    public function filterByContact($contact): self
```

File: src/Centreon/Domain/Contact/Interfaces/ContactFilterInterface.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<?php
+
```
Comment (kduret) [inline src/Centreon/Domain/Contact/Interfaces/ContactFilterInterface.php]: missing header

File: src/Centreon/Domain/Entity/EntityCreator.php [hunk 3]
```diff
@@ -53,8 +53,7 @@ class EntityCreator
      */
     public static function createEntityByArray(string $className, array $data)
```
Comment (kduret) [inline src/Centreon/Domain/Entity/EntityCreator.php]: add return type

File: src/Centreon/Domain/Entity/EntityValidator.php [hunk 4]
```diff
@@ -111,33 +172,75 @@ private function getConstraints(string $entityName, string $groupName, string $t
             // We need to convert camel case to snake case because the data sent
             // are in snake case format whereas the validation definition file
             // use the real name of properties (camel case)
-            $id = $this->convertCamelCaseToSnakeCase($id);
+            $id =self::convertCamelCaseToSnakeCase($id);
```
Comment (kduret) [inline src/Centreon/Domain/Entity/EntityValidator.php]:
```suggestion
            $id = self::convertCamelCaseToSnakeCase($id);
```

File: src/Centreon/Domain/Entity/EntityValidator.php [hunk 5]
```diff
@@ -111,33 +172,75 @@ private function getConstraints(string $entityName, string $groupName, string $t
...
+        }
+    }
+
+    /**
+     * @param ConstraintViolationListInterface $violations
+     * @return ConstraintViolationListInterface
+     */
+    private function removeDuplicatedViolation(
+        ConstraintViolationListInterface $violations
+    ) :ConstraintViolationListInterface {
```
Comment (kduret) [inline src/Centreon/Domain/Entity/EntityValidator.php]:
```suggestion
    ): ConstraintViolationListInterface {
```

File: src/Centreon/Domain/Exception/EntityNotFoundException.php [hunk 6]
```diff
@@ -0,0 +1,13 @@
+<?php
+
```
Comment (kduret) [inline src/Centreon/Domain/Exception/EntityNotFoundException.php]: missing header

File: src/Centreon/Domain/Exception/EntityNotFoundException.php [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+namespace Centreon\Domain\Exception;
+
+use Throwable;
+
+class EntityNotFoundException extends \Exception
+{
+    public function __construct($message = "", Throwable $previous = null)
```
Comment (kduret) [inline src/Centreon/Domain/Exception/EntityNotFoundException.php]: prototype is not the same than the parent ?

File: src/Centreon/Domain/Entity/EntityCreator.php (hunk 3 above)
Comment (callapa) [reply src/Centreon/Domain/Entity/EntityCreator.php]: PHP does not allow to declare a return value as an 'object'. This method is a factory for all object entities. It is therefore not possible to specify a return value. The comment states a mixed return value.

File: src/Centreon/Domain/Exception/EntityNotFoundException.php (hunk 7 above)
Comment (callapa) [reply src/Centreon/Domain/Exception/EntityNotFoundException.php]: Yes, this error code is automatically set to 404.

File: src/Centreon/Domain/Entity/EntityValidator.php [hunk 8]
```diff
@@ -111,33 +172,75 @@ private function getConstraints(string $entityName, string $groupName, string $t
...
+            } else {
+                $violations->remove($index);
+            }
+        }
+        return $violations;
     }
 
     /**
-     * Find the 'Type' contraint from the contraints list.
+     * Find the Type' constraint from the constraints list.
```
Comment (sc979) [inline src/Centreon/Domain/Entity/EntityValidator.php]:
```suggestion
     * Find the 'Type' constraint from the constraints list.
```

File: src/Centreon/Domain/Entity/EntityValidator.php [hunk 9]
```diff
@@ -150,13 +253,49 @@ private function findTypeConstraint(array $constraints): ?string
         return null;
     }
 
+    /**
+     * Formats errors to be more readable.
```
Comment (sc979) [inline src/Centreon/Domain/Entity/EntityValidator.php]:
```suggestion
     * Format errors to be more readable.
```

File: src/Centreon/Domain/Service/AbstractCentreonService.php [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+     * If the filter is defined, all requests will use the ACL of the contact
+     * to fetch data.
+     *
+     * @param mixed $contact Contact to use as a ACL filter
+     * @return self
+     * @throws \Exception
+     */
+    public function filterByContact($contact)
+    {
+        if (!\is_object($contact) || !($contact instanceof ContactInterface)) {
```
Comment (sc979) [inline src/Centreon/Domain/Service/AbstractCentreonService.php]:
```suggestion
        if (!is_object($contact) || !($contact instanceof ContactInterface)) {
```

File: src/Centreon/Infrastructure/RequestParameters/RequestParametersTranslatorException.php [hunk 11]
```diff
@@ -0,0 +1,48 @@
+<?php
+/**
+ * Copyright 2005-2019 Centreon
```
Comment (sc979) [inline src/Centreon/Infrastructure/RequestParameters/RequestParametersTranslatorException.php]: should be apache licence

**Pull request merged**